### PR TITLE
feat (config manager): add bootstrapping_committee

### PR DESCRIPTION
**Pull request by girazoki**
This PR adds a bootstrapping committee in charge of signing the first superblocks until collateral_age plus activity_period epoch is reached.

### CODE REVIEW

File: data_structures/src/superblock.rs [hunk 1]
```diff
@@ -422,19 +435,19 @@ mod tests {
     #[test]
     fn superblock_state_default_add_votes() {
         // When the superblock state is initialized to default (for example when starting the node),
-        // all the received superblock votes are marked as `MaybeValid` or `AlreadySeen`
+        // all the received superblock votes are marked as `NotInArs'
```
Comment (aesedepece) [inline data_structures/src/superblock.rs:438]:
```suggestion
        // all the received superblock votes are marked as `NotInArs`
```

File: node/src/actors/chain_manager/mod.rs [hunk 2]
```diff
@@ -970,7 +970,25 @@ impl ChainManager {
...
         .map_err(|e, _, _| log::error!("Superblock building failed: {:?}", e))
         .and_then(move |(block_headers, last_hash), act, _ctx| {
-            let ars_members = &act.chain_state.last_ars;
+            let chain_info = act.chain_state.chain_info.as_ref().unwrap();
+            let x;
+            let ars_members = {
+                // Before reaching the epoch activity_period + collateral_age the bootstrap committee signs the superblock
+                if act.current_epoch.unwrap()
+                    > chain_info.consensus_constants.collateral_age
+                        + chain_info.consensus_constants.activity_period
```
Comment (aesedepece) [inline node/src/actors/chain_manager/mod.rs:979]: `collateral_age` is measured in blocks, while `activity_period` is measured in epochs. Are we OK to add these together?

File: node/src/actors/chain_manager/mod.rs [hunk 3]
```diff
@@ -970,7 +970,25 @@ impl ChainManager {
...
+                } else {
+                    x = chain_info
+                        .consensus_constants
+                        .bootstrapping_committee
+                        .iter()
+                        .map(|add| add.parse().unwrap())
+                        .collect();
+                    &x
+                }
+            };
```
Comment (aesedepece) [inline node/src/actors/chain_manager/mod.rs:991]: By making `ars_members` not a reference (but referencing later when using it as an argument), we can get rid of the `x` binding altogether.

File: node/src/actors/chain_manager/mod.rs (hunk 2 above)
Comment (lrubiorod) [reply node/src/actors/chain_manager/mod.rs:980]: The idea is to leave decentralized when we are sure that the ARS doesnt contain shitty nodes. So the idea was to wait until requests are living for 1 activity period (stressed net). But as Adan said, if there are empty epochs, the collateral age is not the same, because it is counted in blocks, so should we use an offset?

Comment (girazoki) [reply node/src/actors/chain_manager/mod.rs:980]: I see your concern with a small ARS by the time this timeframe elapses. But was not it the reason why we decided to put this timeframe in the first place? Perhaps we can double it if we feel safer

File: data_structures/src/superblock.rs [hunk 4]
```diff
@@ -57,6 +57,18 @@ pub struct SuperBlockState {
 }
 
 impl SuperBlockState {
+    // Initialize the superblock state
+    pub fn new(superblock_genesis_hash: Hash, bootsrap_committee: Vec<PublicKeyHash>) -> Self {
+        let mut hs = HashSet::new();
+        hs.extend(bootsrap_committee.iter().cloned());
```
Comment (aesedepece) [inline data_structures/src/superblock.rs:63]:
```suggestion
    pub fn new(superblock_genesis_hash: Hash, bootstrap_committee: Vec<PublicKeyHash>) -> Self {
        let hs: HashSet<PublicKeyHash> = bootsrap_committee.iter().cloned().collect();
```

File: data_structures/src/superblock.rs [hunk 5]
```diff
@@ -422,19 +431,19 @@ mod tests {
     #[test]
     fn superblock_state_default_add_votes() {
         // When the superblock state is initialized to default (for example when starting the node),
-        // all the received superblock votes are marked as `MaybeValid` or `AlreadySeen`
+        // all the received superblock votes are marked as `NotInArs`
         let mut sbs = SuperBlockState::default();
```
Comment (tmpolaczyk) [inline data_structures/src/superblock.rs:435]: You could add a test similar to this one but using a non-empty bootstrapping committee. Check that for the first superblock only votes from members of the committee are accepted.

File: node/src/actors/chain_manager/actor.rs [hunk 6]
```diff
@@ -175,11 +176,20 @@ impl ChainManager {
                             },
                         };
 
+                        let bootstrap_committee = chain_info
+                            .consensus_constants
+                            .bootstrapping_committee
+                            .iter()
+                            .map(|add| add.parse().unwrap())
```
Comment (tmpolaczyk) [inline node/src/actors/chain_manager/actor.rs:183]:
```suggestion
                            .map(|add| add.parse().expect("Malformed bootstrapping committee"))
```

File: data_structures/src/superblock.rs (hunk 5 above)
Comment (girazoki) [reply data_structures/src/superblock.rs:435]: great, doing that now

File: node/src/actors/chain_manager/actor.rs [hunk 7]
```diff
@@ -12,6 +12,7 @@ use crate::{
     config_mngr, signature_mngr, storage_mngr,
 };
 use witnet_crypto::key::CryptoEngine;
+use witnet_data_structures::superblock::SuperBlockState;
```
Comment (tmpolaczyk) [inline node/src/actors/chain_manager/actor.rs:15]: fmt

File: node/src/actors/chain_manager/mod.rs [hunk 8]
```diff
@@ -970,12 +963,30 @@ impl ChainManager {
...
+                    > chain_info.consensus_constants.collateral_age
+                        + chain_info.consensus_constants.activity_period
+                {
+                    act.chain_state.last_ars.clone()
+                } else {
+                    chain_info
+                        .consensus_constants
+                        .bootstrapping_committee
+                        .iter()
+                        .map(|add| add.parse().unwrap())
```
Comment (tmpolaczyk) [inline node/src/actors/chain_manager/mod.rs:981]:
```suggestion
                            .map(|add| add.parse().expect("Malformed bootstrapping committee"))
```

File: node/src/actors/chain_manager/mod.rs [hunk 9]
```diff
@@ -1657,7 +1668,7 @@ mod tests {
             chain_manager.get_superblock_beacon(),
             CheckpointBeacon {
                 checkpoint: 0,
-                hash_prev_block: genesis_hash,
+                hash_prev_block: Hash::default(),
```
Comment (tmpolaczyk) [inline node/src/actors/chain_manager/mod.rs:1671]: I think this test is supposed to check that the superblock genesis hash is set correctly. And it is not, because the `SuperBlockState` is not initialized. You could try to do `chain_manager.chain_state.superblock_state = SuperBlockState::new(...)` and set the genesis hash there.

File: data_structures/src/superblock.rs [hunk 10]
```diff
@@ -127,26 +142,19 @@ impl SuperBlockState {
...
     fn is_valid(&self, sbv: &SuperBlockVote) -> Option<bool> {
-        match self.current_superblock_index {
-            // We do not know the current index, we cannot know if the vote is valid
-            None => None,
-            // If the index is the same as the current one, the vote is valid if it is signed by a
-            // member of the ARS
-            Some(x) if x == sbv.superblock_index => self
-                .previous_ars_identities
+        if self.current_superblock_index == sbv.superblock_index {
+            self.current_signing_committee
```
Comment (tmpolaczyk) [inline data_structures/src/superblock.rs:146]: Hey, don't remove the comments
```suggestion
            // If the index is the same as the current one, the vote is valid if it is signed by a
            // member of the current signing committee
            self.current_signing_committee
```

File: data_structures/src/superblock.rs [hunk 11]
```diff
@@ -224,13 +228,12 @@ impl SuperBlockState {
...
     }
 
-    /// Returns an option with the last superblock hash and index if none of these fields is None.
-    /// else, returns None
-    pub fn get_beacon(&self) -> Option<CheckpointBeacon> {
-        Some(CheckpointBeacon {
-            checkpoint: self.current_superblock_index?,
-            hash_prev_block: self.current_superblock_hash?,
-        })
+    /// Returns ahe last superblock hash and index.
```
Comment (aesedepece) [inline data_structures/src/superblock.rs:231]:
```suggestion
    /// Returns the last superblock hash and index.
```

File: data_structures/src/superblock.rs [hunk 12]
```diff
@@ -509,9 +546,9 @@ mod tests {
         );
 
         // The only think that should change is the superblock_index
-        expected_sbs.current_superblock_index = Some(1);
-        // And the superblock_hash, which will be set to None
-        expected_sbs.current_superblock_hash = None;
+        expected_sbs.current_superblock_index = 1;
+        // And the superblock_hash, which will be set to the previous superblock
+        expected_sbs.current_superblock_hash = _sb1.hash();
```
Comment (tmpolaczyk) [inline data_structures/src/superblock.rs:551]: You should remove the `_` if the variable is no longer unused.

```suggestion
        expected_sbs.current_superblock_hash = sb1.hash();
```

File: data_structures/src/superblock.rs (hunk 10 above)
Comment (aesedepece) [reply data_structures/src/superblock.rs:148]: Is he?

File: data_structures/src/superblock.rs [hunk 13]
```diff
@@ -127,26 +142,21 @@ impl SuperBlockState {
...
+                .map_or(Some(false), |x| {
+                    Some(x.contains(&sbv.secp256k1_signature.public_key.pkh()))
+                })
+        } else if ((self.current_superblock_index.saturating_sub(1))
+            ..=self.current_superblock_index.saturating_add(1))
+            .contains(&sbv.superblock_index)
+        {
+            None
+        } else {
+            Some(false)
```
Comment (tmpolaczyk) [inline data_structures/src/superblock.rs:159]:
```suggestion
            // If the index is not the same as the current one, but it is within an acceptable range 
            // of [x-1, x+1], broadcast the vote without checking if it is a member of the ARS, as 
            // the ARS may have changed and we do not keep older copies of the ARS in memory 
            None
        } else {
            // If the index is outside the [x-1, x+1] range, it is considered not valid
            Some(false)
```
